code/espurna/sensors/PZEM004TV30Sensor.h

```diff
@@ -0,0 +1,573 @@
+
+        // - Current: 4 bytes, in 0.001A (we return A)
+        _current = take_4();
+        _current /= 1000.0;
+
+        // - Power: 4 bytes, in 0.1W (we return W)
+        _power = take_4();
+        _power /= 10.0;
+
+        // - Energy: 4 bytes, in Wh (we return kWh)
+        _energy = take_4();
+        _energy /= 1000.0;
+
+        // - Frequency: 2 bytes, in 0.1Hz (we return Hz)
+        _frequency = take_2();
+        _frequency /= 10.0;
+
+        // - Power Factor: 2 bytes in 0.1% (we return %)
+        _power_factor = take_2();
+        _power_factor /= 10.0;
+
+        // - Alarms: 2 bytes, (NOT IMPLEMENTED)
+        // XXX: it seems it can only be either 0xffff or 0 for ON and OFF respectively
+        // XXX: what this does, exactly?
+        _alarm = (0xff == *it) && (0xff == *(it + 1));
+    }
+
+    // Reading measurements is a standard modbus function:
+    // - addr, 0x04, rhigh, rlow, rnumhigh, rnumlow, crchigh, crclow
+    // ReadInput reply can be one of:
+    // - addr, 0x04, nbytes, rndatahigh, rndatalow, rndata..., crchigh, crclow (on success)
+    // - addr, 0x84, error_code, crchigh, crclow (on error. modbus rtu sets high bit to 1 i.e. 0b00000100 becomes 0b10000100)
+    void modbusReadValues() {
+        _error = SENSOR_ERROR_OK;
+
+        auto request = adu_builder(_address, ReadInputCode)
+            .add(static_cast<uint16_t>(0))
+            .add(static_cast<uint16_t>(10))
+            .end();
+        modbusProcess(request, [this](buffer_type&& buffer, size_t size) {
+            parseMeasurements(std::move(buffer), size);
+        });
+    }
+
+    void flush() {
+        while (_stream->read() >= 0) {
+        }
+    }
+
+    // ---------------------------------------------------------------------
+
+    // Note that the device (aka slave) address needs be changed first via
+    // - some external tool. For example, using USB2TTL adapter and a PC app
+    // - `pzem.address` with **only** one device on the line
+    //    (because we would change all 0xf8-addressed devices at the same time)
+    void setAddress(uint8_t address) {
+        _address = address;
+    }
+
+    void setDebug(bool debug) {
+        _debug = debug;
+    }
+
+    void setStream(Stream* stream) {
+        _stream = stream;
+        _stream->setTimeout(_read_timeout);
+    }
+
+    void setReadTimeout(unsigned long value) {
+        _read_timeout = value;
+    }
+
+    void setUpdateInterval(unsigned long value) {
+        _update_interval = value;
+    }
+
+    template <typename T>
+    void setDescription(T&& description) {
+        _description = std::forward<T>(description);
+    }
+
+    // ---------------------------------------------------------------------
+
+    void begin() override {
+        _ready = (_stream != nullptr);
+        _last_reading = millis() - _update_interval;
+        #if TERMINAL_SUPPORT
+            terminalRegisterCommand(F("PZ.ADDRESS"), [](const terminal::CommandContext& ctx) {
```

Comment:
`pz.address N` should change PZEM's address to N, where N is some number between 1 and 247 (per http://www.modbus.org/docs/Modbus_over_serial_line_V1_01.pdf)
PZEM pdf never mentioned 0 as broadcast, only 248 (0xF8), so it is unclear what exactly happens with numbers 0,248...255.